File: y2024/day12/day12.py
import os
import traceback
from enum import Enum, auto
from pprint import pprint

SAMPLE = """
AAAA
BBCD
BBCC
EEEC
"""
SAMPLE = """
RRRRIICCFF
RRRRIICCCF
VVRRRCCFFF
VVRCCCJFFF
VVVVCJJCFE
VVIVCCJJEE
VVIIICJJEE
MIIIIIJJEE
MIIISIJEEE
MMMISSJEEE
"""


class Direction(Enum):
    up = auto()
    down = auto()
    left = auto()
    right = auto()


def process_input():
    with open("./day12/input.txt", "r") as fp:
        content = fp.readlines()
    # content = SAMPLE.splitlines()
    content = [list(i.strip()) for i in content if i != ""]
    return content


def count_edges(
    positions: list[list[str]],
    row: int,
    col: int,
    curr_plant: str,
    rowlim: int,
    collim: int,
) -> int:
    edges = 0
    # up
    try:
        if row == 0:
            edges += 1
        elif row - 1 >= 0 and positions[row - 1][col] != curr_plant:
            edges += 1
        # down
        if row == rowlim - 1:
            edges += 1
        elif row + 1 < rowlim and positions[row + 1][col] != curr_plant:
            edges += 1
        # right
        if col == collim - 1:
            edges += 1
        elif col + 1 < collim and positions[row][col + 1] != curr_plant:
            edges += 1
        # left
        if col == 0:
            edges += 1
        elif col - 1 >= 0 and positions[row][col - 1] != curr_plant:
            edges += 1
    except IndexError:
        print(f"index error at: {row=}, {col=}")
        print(traceback.format_exc())
        os._exit(1)
    return edges


def get_edges(
    positions: list[list[str]],
    coord: tuple[int, int],
    edge_tracker: dict[tuple[int, int], int],
    visited_coords: set,
    row_lim: int,
    col_lim: int,
    # plant_nodes: dict[str, list[tuple[int, int]]]
) -> tuple[int, int]:
    if coord in visited_coords:
        return 0, 0
    visited_coords.add(coord)
    curr_plant = positions[coord[0]][coord[1]]
    # print(f"\n [X] {curr_plant} - adding coord: {coord}")
    num_edges = edge_tracker[coord]
    total_area = 1
    if num_edges == 4:
        return num_edges, total_area

    # can you go up:
    if coord[0] - 1 >= 0 and positions[coord[0] - 1][coord[1]] == curr_plant:
        edge, area = get_edges(
            positions=positions,
            coord=(coord[0] - 1, coord[1]),
            edge_tracker=edge_tracker,
            visited_coords=visited_coords,
            row_lim=row_lim,
            col_lim=col_lim,
        )
        num_edges += edge
        total_area += area
    # can you go down:
    if coord[0] + 1 < row_lim and positions[coord[0] + 1][coord[1]] == curr_plant:
        edge, area = get_edges(
            positions,
            (coord[0] + 1, coord[1]),
            edge_tracker,
            visited_coords,
            row_lim,
            col_lim,
        )
        num_edges += edge
        total_area += area
    # can you go left:
    if coord[1] - 1 >= 0 and positions[coord[0]][coord[1] - 1] == curr_plant:
        edge, area = get_edges(
            positions,
            (coord[0], coord[1] - 1),
            edge_tracker,
            visited_coords,
            row_lim,
            col_lim,
        )
        num_edges += edge
        total_area += area
    # can you go right:
    if coord[1] + 1 < col_lim and positions[coord[0]][coord[1] + 1] == curr_plant:
        edge, area = get_edges(
            positions,
            (coord[0], coord[1] + 1),
            edge_tracker,
            visited_coords,
            row_lim,
            col_lim,
        )
        num_edges += edge
        total_area += area

    return num_edges, total_area


def make_edge_count(
    positions: list[list[str]], rowlim: int, collim: int
) -> dict[tuple[int, int], int]:
    edge_map = {}
    for row_num, row in enumerate(positions):
        for col_num, col in enumerate(row):
            coordinate = (row_num, col_num)
            edge_map[coordinate] = count_edges(
                positions, row_num, col_num, col, rowlim, collim
            )
    return edge_map


def part01(data: list[list[str]]):
    visited_plant_coordinates: set = set()
    row_lim = len(data)
    col_lim = len(data[0])
    edge_tracker: dict[tuple[int, int], int] = make_edge_count(data, row_lim, col_lim)
    price = 0
    for row_num, row in enumerate(data):
        for col_num, plant in enumerate(row):
            coord = (row_num, col_num)
            if coord in visited_plant_coordinates:
                continue
            perimeter, area = get_edges(
                positions=data,
                coord=coord,
                edge_tracker=edge_tracker,
                visited_coords=visited_plant_coordinates,
                row_lim=row_lim,
                col_lim=col_lim,
            )
            price += perimeter * area
            # print(f"{plant=}, {perimeter=}, {area=}")

    print(f"price: {price}")


def get_continuous_edges(
    positions: list[list[str]],
    coord: tuple[int, int],
    edge_tracker: dict[tuple[int, int], int],
    visited_coords: set,
    row_lim: int,
    col_lim: int,
    positional_coords: set,
) -> tuple[int, int]:
    if coord in visited_coords:
        return 0, 0
    visited_coords.add(coord)
    curr_plant = positions[coord[0]][coord[1]]
    # print(f"\n [X] {curr_plant} - adding coord: {coord}")
    num_edges = edge_tracker[coord]
    if num_edges == 0:
        return 0, 1
    total_area = 1
    if num_edges == 4:
        return num_edges, total_area

    # can you go up:
    if coord[0] - 1 >= 0 and positions[coord[0] - 1][coord[1]] == curr_plant:
        edge, area = get_edges(
            positions=positions,
            coord=(coord[0] - 1, coord[1]),
            edge_tracker=edge_tracker,
            visited_coords=visited_coords,
            row_lim=row_lim,
            col_lim=col_lim,
        )
        num_edges += edge
        total_area += area
    # can you go down:
    if coord[0] + 1 < row_lim and positions[coord[0] + 1][coord[1]] == curr_plant:
        edge, area = get_edges(
            positions,
            (coord[0] + 1, coord[1]),
            edge_tracker,
            visited_coords,
            row_lim,
            col_lim,
        )
        num_edges += edge
        total_area += area
    # can you go left:
    if coord[1] - 1 >= 0 and positions[coord[0]][coord[1] - 1] == curr_plant:
        edge, area = get_edges(
            positions,
            (coord[0], coord[1] - 1),
            edge_tracker,
            visited_coords,
            row_lim,
            col_lim,
        )
        num_edges += edge
        total_area += area
    # can you go right:
    if coord[1] + 1 < col_lim and positions[coord[0]][coord[1] + 1] == curr_plant:
        edge, area = get_edges(
            positions,
            (coord[0], coord[1] + 1),
            edge_tracker,
            visited_coords,
            row_lim,
            col_lim,
        )
        num_edges += edge
        total_area += area

    return num_edges, total_area


def part02(data: list[list[str]]):
    visited_plant_coordinates: set = set()
    row_lim = len(data)
    col_lim = len(data[0])
    edge_tracker: dict[tuple[int, int], int] = make_edge_count(data, row_lim, col_lim)
    price = 0
    letter_tracker: dict = {}
    for row_num, row in enumerate(data):
        for col_num, col in enumerate(row):
            positional_coords = set()
            area = get_continuous_edges(
                positions=data,
                coord=(row_num, col_num),
                edge_tracker=edge_tracker,
                visited_coords=visited_plant_coordinates,
                row_lim=row_lim,
                col_lim=col_lim,
                positional_coords=positional_coords,
            )


def main():
    import time

    data = process_input()
    # pprint(data)
    start = time.time()
    part01(data)
    print(f"time taken part 01: {(time.time() - start)*1000}")
    part02(data)


if __name__ == "__main__":
    main()

File: y2024/day12/day12.go
package day12

import (
	"adventofcode/pkg/graph"
	"adventofcode/y2024/utils"
	"fmt"
)

const SAMPLE = `RRRRIICCFF
RRRRIICCCF
VVRRRCCFFF
VVRCCCJFFF
VVVVCJJCFE
VVIVCCJJEE
VVIIICJJEE
MIIIIIJJEE
MIIISIJEEE
MMMISSJEEE`

func processInput(daydir string) [][]rune {
	lines := utils.ReadLines(daydir + "/input.txt")
	// lines := strings.Split(SAMPLE, "\n")
	retval := [][]rune{}
	for _, l := range lines {
		retval = append(retval, []rune(l))
	}
	return retval
}

func countEdges(grid [][]rune, row, col, rowLim, colLim int, currPlant rune) int {
	edges := 0
	// up
	if row == 0 {
		edges++
	} else if row-1 >= 0 && grid[row-1][col] != currPlant {
		edges++
	}
	// down
	if row == rowLim-1 {
		edges++
	} else if row+1 < rowLim && grid[row+1][col] != currPlant {
		edges++
	}
	// left
	if col == 0 {
		edges++
	} else if col-1 >= 0 && grid[row][col-1] != currPlant {
		edges++
	}
	// right
	if col == colLim-1 {
		edges++
	} else if col+1 <= colLim && grid[row][col+1] != currPlant {
		edges++
	}

	return edges
}

func getEdges(
	grid [][]rune, coord graph.Coordinate, rowLim, colLim int,
	edgeTracker map[graph.Coordinate]int, visitedCoords map[graph.Coordinate]bool,
) (int, int) {
	if _, ok := visitedCoords[coord]; ok {
		return 0, 0
	}
	visitedCoords[coord] = true
	currPlant := grid[coord.Row][coord.Col]
	numEdges := edgeTracker[coord]
	totalArea := 1
	if numEdges == 4 {
		return numEdges, totalArea
	}

	// up
	if coord.Row-1 >= 0 && grid[coord.Row-1][coord.Col] == currPlant {
		edge, area := getEdges(
			grid,
			graph.Coordinate{Row: coord.Row - 1, Col: coord.Col},
			rowLim, colLim, edgeTracker, visitedCoords,
		)
		numEdges += edge
		totalArea += area
	}
	// down
	if coord.Row+1 < rowLim && grid[coord.Row+1][coord.Col] == currPlant {
		edge, area := getEdges(
			grid,
			graph.Coordinate{Row: coord.Row + 1, Col: coord.Col},
			rowLim, colLim, edgeTracker, visitedCoords,
		)
		numEdges += edge
		totalArea += area
	}
	// left
	if coord.Col-1 >= 0 && grid[coord.Row][coord.Col-1] == currPlant {
		edge, area := getEdges(
			grid,
			graph.Coordinate{Row: coord.Row, Col: coord.Col - 1},
			rowLim, colLim, edgeTracker, visitedCoords,
		)
		numEdges += edge
		totalArea += area
	}
	// right
	if coord.Col+1 < colLim && grid[coord.Row][coord.Col+1] == currPlant {
		edge, area := getEdges(
			grid,
			graph.Coordinate{Row: coord.Row, Col: coord.Col + 1},
			rowLim, colLim, edgeTracker, visitedCoords,
		)
		numEdges += edge
		totalArea += area
	}

	return numEdges, totalArea
}

func makeEdgeTracker(grid [][]rune, rowLim, colLim int) map[graph.Coordinate]int {
	edgeMap := map[graph.Coordinate]int{}
	for rowNum, row := range grid {
		for colNum, val := range row {
			coord := graph.Coordinate{Row: rowNum, Col: colNum}
			edgeMap[coord] = countEdges(grid, rowNum, colNum, rowLim, colLim, val)
		}
	}
	return edgeMap
}

func part01(grid [][]rune) {
	visitedPlantCoordinates := map[graph.Coordinate]bool{}
	rowLim := len(grid)
	colLim := len(grid[0])
	edgeTracker := makeEdgeTracker(grid, rowLim, colLim)
	price := 0
	for rowNum, row := range grid {
		for colNum := range len(row) {
			coord := graph.Coordinate{Row: rowNum, Col: colNum}
			if _, ok := visitedPlantCoordinates[coord]; ok {
				continue
			}
			perimeter, area := getEdges(grid, coord, rowLim, colLim, edgeTracker, visitedPlantCoordinates)
			price += perimeter * area
		}
	}
	fmt.Println("PART 01:", price)
}

func Run(dir string) {
	utils.PartPrinter("DAY 12")
	grid := processInput(dir + "/day12")
	part01(grid)
}
